File: internal/domain/surcharge.go
package domain

// GetTypeSurcharge returns the surcharge for a specific item type.
// For standard types, uses job's type-specific fields.
// For custom types, uses the custom type's surcharge or job default.
func GetTypeSurcharge(job *Job, itemType LineItemType, customTypes []*JobItemType) float64 {
	switch itemType {
	case LineItemTypeMaterial:
		if job.MaterialSurchargePercent != nil {
			return *job.MaterialSurchargePercent
		}
	case LineItemTypeLabor:
		if job.LaborSurchargePercent != nil {
			return *job.LaborSurchargePercent
		}
	case LineItemTypeEquipment:
		if job.EquipmentSurchargePercent != nil {
			return *job.EquipmentSurchargePercent
		}
	default:
		// Custom type - look up in customTypes
		for _, ct := range customTypes {
			if ct.Slug == string(itemType) && ct.SurchargePercent != nil {
				return *ct.SurchargePercent
			}
		}
	}
	// Fall back to job default
	return job.SurchargePercent
}

// EffectiveSurcharge calculates the applicable surcharge for a line item
// based on the job's surcharge mode, item type, and the category hierarchy.
func EffectiveSurcharge(li *LineItem, job *Job, categoryChain []*Category, customTypes []*JobItemType) float64 {
	if job.SurchargeMode == SurchargeModeOverride {
		return effectiveSurchargeOverride(li, job, categoryChain, customTypes)
	}
	return effectiveSurchargeStacking(li, job, categoryChain, customTypes)
}

// effectiveSurchargeOverride returns the most specific (lowest-level) surcharge.
// Priority: LineItem > deepest Category > Type surcharge
func effectiveSurchargeOverride(li *LineItem, job *Job, categoryChain []*Category, customTypes []*JobItemType) float64 {
	// Check line item first
	if li.SurchargePercent != nil {
		return *li.SurchargePercent
	}

	// Walk category chain from deepest to shallowest
	for i := len(categoryChain) - 1; i >= 0; i-- {
		if categoryChain[i].SurchargePercent != nil {
			return *categoryChain[i].SurchargePercent
		}
	}

	// Fall back to type-specific surcharge
	return GetTypeSurcharge(job, li.Type, customTypes)
}

// effectiveSurchargeStacking sums all surcharges in the hierarchy.
// Total = Type% + Category%s + LineItem%
func effectiveSurchargeStacking(li *LineItem, job *Job, categoryChain []*Category, customTypes []*JobItemType) float64 {
	// Start with type-specific surcharge instead of job.SurchargePercent
	total := GetTypeSurcharge(job, li.Type, customTypes)

	// Add all category surcharges
	for _, cat := range categoryChain {
		if cat.SurchargePercent != nil {
			total += *cat.SurchargePercent
		}
	}

	// Add line item surcharge
	if li.SurchargePercent != nil {
		total += *li.SurchargePercent
	}

	return total
}

// FinalPrice calculates the line item total with surcharge applied.
func FinalPrice(li *LineItem, effectiveSurcharge float64) float64 {
	base := li.BasePrice()
	return base * (1 + effectiveSurcharge/100)
}

// CategoryTotal calculates the total for a category including all line items and child categories.
type CategoryTotal struct {
	CategoryID     string  `json:"category_id"`
	Subtotal       float64 `json:"subtotal"`        // Sum of base prices
	SurchargeTotal float64 `json:"surcharge_total"` // Sum of surcharges
	Total          float64 `json:"total"`           // Final total
}

// JobTotal calculates the complete job totals.
type JobTotal struct {
	Subtotal       float64            `json:"subtotal"`        // Sum of all base prices
	SurchargeTotal float64            `json:"surcharge_total"` // Total surcharges applied
	GrandTotal     float64            `json:"grand_total"`     // Final total
	TypeSubtotals  map[string]float64 `json:"type_subtotals"`  // Subtotals keyed by type slug
}

// MaterialSubtotal returns the subtotal for material items (backward compatibility).
func (jt JobTotal) MaterialSubtotal() float64 {
	return jt.TypeSubtotals["material"]
}

// LaborSubtotal returns the subtotal for labor items (backward compatibility).
func (jt JobTotal) LaborSubtotal() float64 {
	return jt.TypeSubtotals["labor"]
}

// EquipmentSubtotal returns the subtotal for equipment items (backward compatibility).
func (jt JobTotal) EquipmentSubtotal() float64 {
	return jt.TypeSubtotals["equipment"]
}

// CalculateJobTotal computes all totals for a job.
func CalculateJobTotal(job *Job, categories []*Category, lineItems []*LineItem, customTypes []*JobItemType) JobTotal {
	var result JobTotal
	result.TypeSubtotals = make(map[string]float64)

	// Build category lookup for chain resolution
	categoryByID := make(map[string]*Category)
	for _, cat := range categories {
		categoryByID[cat.ID] = cat
	}

	// Build category chain for each line item's category
	categoryChains := make(map[string][]*Category)

	for _, li := range lineItems {
		// Get or build category chain
		chain, exists := categoryChains[li.CategoryID]
		if !exists {
			chain = buildCategoryChain(li.CategoryID, categoryByID)
			categoryChains[li.CategoryID] = chain
		}

		// Calculate effective surcharge and prices
		basePrice := li.BasePrice()
		effSurcharge := EffectiveSurcharge(li, job, chain, customTypes)
		finalPrice := FinalPrice(li, effSurcharge)

		result.Subtotal += basePrice
		result.GrandTotal += finalPrice

		// Track by type using dynamic map
		typeKey := string(li.Type)
		result.TypeSubtotals[typeKey] += finalPrice
	}

	result.SurchargeTotal = result.GrandTotal - result.Subtotal

	return result
}

// buildCategoryChain builds the chain from root to the specified category.
func buildCategoryChain(categoryID string, categoryByID map[string]*Category) []*Category {
	var chain []*Category
	current := categoryByID[categoryID]

	for current != nil {
		chain = append([]*Category{current}, chain...) // Prepend to build root-first
		if current.ParentID == nil {
			break
		}
		current = categoryByID[*current.ParentID]
	}

	return chain
}

// CalculateCategoryTotal computes totals for a category including all nested subcategories.
func CalculateCategoryTotal(categoryID string, job *Job, categories []*Category, lineItems []*LineItem, customTypes []*JobItemType) CategoryTotal {
	var result CategoryTotal
	result.CategoryID = categoryID

	// Build category lookup
	categoryByID := make(map[string]*Category)
	for _, cat := range categories {
		categoryByID[cat.ID] = cat
	}

	// Find all descendant category IDs (including the target category itself)
	descendantIDs := findDescendantCategories(categoryID, categories)
	descendantIDs[categoryID] = true

	// Build category chains cache
	categoryChains := make(map[string][]*Category)

	for _, li := range lineItems {
		// Only include items from this category or its descendants
		if !descendantIDs[li.CategoryID] {
			continue
		}

		// Get or build category chain
		chain, exists := categoryChains[li.CategoryID]
		if !exists {
			chain = buildCategoryChain(li.CategoryID, categoryByID)
			categoryChains[li.CategoryID] = chain
		}

		// Calculate prices
		basePrice := li.BasePrice()
		effSurcharge := EffectiveSurcharge(li, job, chain, customTypes)
		finalPrice := FinalPrice(li, effSurcharge)

		result.Subtotal += basePrice
		result.Total += finalPrice
	}

	result.SurchargeTotal = result.Total - result.Subtotal

	return result
}

// findDescendantCategories returns a set of all category IDs that are descendants of the given category.
func findDescendantCategories(parentID string, categories []*Category) map[string]bool {
	result := make(map[string]bool)

	// Build children lookup
	childrenOf := make(map[string][]string)
	for _, cat := range categories {
		if cat.ParentID != nil {
			childrenOf[*cat.ParentID] = append(childrenOf[*cat.ParentID], cat.ID)
		}
	}

	// BFS to find all descendants
	queue := childrenOf[parentID]
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		result[current] = true
		queue = append(queue, childrenOf[current]...)
	}

	return result
}

File: frontend/src/surcharge.ts
// Client-side replication of internal/domain/surcharge.go
// Must produce identical results to the Go domain layer.

import type { QuoteJob, QuoteCategory, QuoteLineItem, QuoteItemType, QuoteTotals } from './types';

/** Get the surcharge for a specific item type (mirrors domain.GetTypeSurcharge). */
export function getTypeSurcharge(
  job: QuoteJob,
  itemType: string,
  customTypes: QuoteItemType[],
): number {
  switch (itemType) {
    case 'material':
      if (job.material_surcharge_percent != null) return job.material_surcharge_percent;
      break;
    case 'labor':
      if (job.labor_surcharge_percent != null) return job.labor_surcharge_percent;
      break;
    case 'equipment':
      if (job.equipment_surcharge_percent != null) return job.equipment_surcharge_percent;
      break;
    default:
      for (const ct of customTypes) {
        if (ct.slug === itemType && ct.surcharge_percent != null) {
          return ct.surcharge_percent;
        }
      }
  }
  return job.surcharge_percent;
}

/** Build category chain from root to target (mirrors domain.buildCategoryChain). */
function buildCategoryChain(
  categoryId: string,
  categoryById: Map<string, QuoteCategory>,
): QuoteCategory[] {
  const chain: QuoteCategory[] = [];
  let current = categoryById.get(categoryId);
  while (current) {
    chain.unshift(current);
    if (current.parent_id == null) break;
    current = categoryById.get(current.parent_id);
  }
  return chain;
}

/** Effective surcharge in override mode (most specific wins). */
function effectiveSurchargeOverride(
  item: QuoteLineItem,
  job: QuoteJob,
  chain: QuoteCategory[],
  customTypes: QuoteItemType[],
): number {
  if (item.surcharge_percent != null) return item.surcharge_percent;
  for (let i = chain.length - 1; i >= 0; i--) {
    if (chain[i].surcharge_percent != null) return chain[i].surcharge_percent!;
  }
  return getTypeSurcharge(job, item.type, customTypes);
}

/** Effective surcharge in stacking mode (all add together). */
function effectiveSurchargeStacking(
  item: QuoteLineItem,
  job: QuoteJob,
  chain: QuoteCategory[],
  customTypes: QuoteItemType[],
): number {
  let total = getTypeSurcharge(job, item.type, customTypes);
  for (const cat of chain) {
    if (cat.surcharge_percent != null) total += cat.surcharge_percent;
  }
  if (item.surcharge_percent != null) total += item.surcharge_percent;
  return total;
}

/** Calculate effective surcharge for a line item (mirrors domain.EffectiveSurcharge). */
export function effectiveSurcharge(
  item: QuoteLineItem,
  job: QuoteJob,
  chain: QuoteCategory[],
  customTypes: QuoteItemType[],
): number {
  if (job.surcharge_mode === 'override') {
    return effectiveSurchargeOverride(item, job, chain, customTypes);
  }
  return effectiveSurchargeStacking(item, job, chain, customTypes);
}

/** Flatten all categories into a map by ID. */
function flattenCategories(categories: QuoteCategory[]): Map<string, QuoteCategory> {
  const map = new Map<string, QuoteCategory>();
  function walk(cats: QuoteCategory[]) {
    for (const cat of cats) {
      map.set(cat.id, cat);
      walk(cat.children);
    }
  }
  walk(categories);
  return map;
}

/** Collect all line items from nested category tree. */
function collectAllItems(categories: QuoteCategory[]): QuoteLineItem[] {
  const items: QuoteLineItem[] = [];
  function walk(cats: QuoteCategory[]) {
    for (const cat of cats) {
      items.push(...cat.items);
      walk(cat.children);
    }
  }
  walk(categories);
  return items;
}

/** Calculate full job totals (mirrors domain.CalculateJobTotal). */
export function calculateTotals(
  job: QuoteJob,
  categories: QuoteCategory[],
  customTypes: QuoteItemType[],
): QuoteTotals {
  const categoryById = flattenCategories(categories);
  const chainCache = new Map<string, QuoteCategory[]>();
  const allItems = collectAllItems(categories);

  let subtotal = 0;
  let grandTotal = 0;
  const typeSubtotals: Record<string, number> = {};

  for (const item of allItems) {
    let chain = chainCache.get(item.category_id);
    if (!chain) {
      chain = buildCategoryChain(item.category_id, categoryById);
      chainCache.set(item.category_id, chain);
    }

    const basePrice = item.quantity * item.unit_price;
    const effSurcharge = effectiveSurcharge(item, job, chain, customTypes);
    const finalPrice = basePrice * (1 + effSurcharge / 100);

    subtotal += basePrice;
    grandTotal += finalPrice;
    typeSubtotals[item.type] = (typeSubtotals[item.type] || 0) + finalPrice;
  }

  return {
    subtotal,
    surcharge_total: grandTotal - subtotal,
    grand_total: grandTotal,
    type_subtotals: typeSubtotals,
  };
}
